Histfit doc

jonas-kleck opened the pull request:
Added documentation for Histogram Fits. Added an example that shows, why we can't fit a histogram using XYFit so easily, so why to use HistFit for histogrammed data, since @GuenterQuast suggested that histogram fits are often problematic for students. I am a bit unsure about the example, should I include more/other problems?

jonas-kleck committed:
added documentation for histogram fits, added an example showing why to use histogram fits

jonas-kleck committed:
edited Histogram Fit section

GuenterQuast commented:
Thank you for addressing this important issue !

In addition to the problems of an xy-fit, which you mention correctly, 
there is another one:
- tanking the uncertainty as sqrt(n_i_observed) in a bin leads to a 
bias, because downward-
fluctuations are assigned smaller uncertainties than 
upward-fluctuations. As a remedy,
the uncertainties may be taken from the prediction of a pre-fit, if one 
insists to use
an xy-fit. Using the Poisson-Likelihood of histfit also fixes this issue.

You man want to address and illustrate this point as well ?


Am 18.09.25 um 11:45 schrieb jonas-kleck:
>
> Added documentation for Histogram Fits. Added an example that shows, 
> why we can't fit a histogram using XYFit so easily, so why to use 
> HistFit for histogrammed data, since @GuenterQuast 
> <https://github.com/GuenterQuast> suggested that histogram fits are 
> often problematic for students. I am a bit unsure about the example, 
> should I include more/other problems?
>
> ------------------------------------------------------------------------
>
>
>         You can view, comment on, or merge this pull request online at:
>
> https://github.com/PhiLFitters/kafe2/pull/258
>
>
>         Commit Summary
>
>   * 29d82ea
>     <https://github.com/PhiLFitters/kafe2/pull/258/commits/29d82ea60f59786ef7154eea0f5d9465bcb90f36>
>     added documentation for histogram fits, added an example showing
>     why to use histogram fits
>   * 3dcaf2c
>     <https://github.com/PhiLFitters/kafe2/pull/258/commits/3dcaf2c4fb1c2dea686eb3c3806fb50010cb5783>
>     edited Histogram Fit section
>
>
>         File Changes
>
> (2 files <https://github.com/PhiLFitters/kafe2/pull/258/files>)
>
>   * *M* doc/src/parts/user_guide.rst
>     <https://github.com/PhiLFitters/kafe2/pull/258/files#diff-3b835a49964ee8e16de7da16729e8034c6cec33ab20ea79760471209d6ce85f2>
>     (23)
>   * *A* examples/009_histogram_fit/04_pitfalls.py
>     <https://github.com/PhiLFitters/kafe2/pull/258/files#diff-7f410467fb771a66e6ab72917a1ea5076ea871ab4605d1397c7c6df1c912c715>
>     (115)
>
>
>         Patch Links:
>
>   * https://github.com/PhiLFitters/kafe2/pull/258.patch
>   * https://github.com/PhiLFitters/kafe2/pull/258.diff
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/PhiLFitters/kafe2/pull/258>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AEIAGSNMHIWKX4EGO4TPD433TJ5MNAVCNFSM6AAAAACG3C4DJCVHI2DSMVQWIX3LMV43ASLTON2WKOZTGQZDSNJUGY3TINY>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

jonas-kleck committed:
Added prefit as solution to the incorrectly specified uncertainties to example

jonas-kleck committed:
corrected documentation

JohannesGaessler commented on doc/src/parts/user_guide.rst:
```diff
@@ -389,6 +389,29 @@ A typical dictionary returned by the :py:meth:`~.FitBase.do_fit` method looks li
+Histogram Fits
+---------------
+
+A very common fit type is the histogram fit. *kafe2* provides a dedicated fitting class for histogram fits,
```
```suggestion
In physics experiments data is frequently histogrammed in order to reduce the data to a manageable number of bins.
*kafe2* provides a dedicated fitting class for histogram fits,
```

JohannesGaessler commented on doc/src/parts/user_guide.rst:
```diff
@@ -389,6 +389,29 @@ A typical dictionary returned by the :py:meth:`~.FitBase.do_fit` method looks li
+    
+Depending on whether the modelfunction is already normalised, or has a normalisation constant,
+that is also supposed to be estimated in the fit the `density` keyword can be used, during creation
+of the `HistFit` object.
```
```suggestion
By default it is assumed that the model function for a `HistFit` object is a probability density function normalised to 1.
As a consequence the bin contents are also being normalized to 1.
To disable this behavior, set `density=False` in the `HistFit` constructor.
```

I think this explains the behavior more explicitly. We can still mention the fitting of normalization constants, though you could in principle also have a model function with a constant normalization that you're not fitting.

JohannesGaessler commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,166 @@
+=======================================
+
+This example demonstrates why it is more convenient to use the HistFit class
+instead of XYFit when dealing with histogrammed data.
```
```suggestion
This example demonstrates a scenario in which it is more convenient to use the
HistFit class rather than the XYFit class.
```

JohannesGaessler commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,166 @@
+
+While it is in principle possible to perform such fits correctly using XYFit,
+it requires much more care. This example shows common mistakes that can occur
+when that necessary care is not taken and how this makes the fit results worse.
```
```suggestion
While it is in principle possible to perform such fits correctly using XYFit,
more care must be taken to avoid unusable or biased results. 
This example shows common problems that can occur when using XYFit and how to fix them.
HistFit handles these problems automatically.
```

JohannesGaessler commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,166 @@
+x_data = binmids = np.mean([binedges[:-1], binedges[1:]], axis=0) #use binmids as x values
+y_data = bincounts/(np.sum(bincounts)*np.diff(binedges))
+print(y_data) #use normalized histogram as y data
+x_error = 0.25 #use half binwidht as x_error
```
```suggestion
x_error = 0.25 #use half binwidth as x_error
```

cverstege replied:
```suggestion
x_error = 0.25 #use half binwidth as x_error
```

JohannesGaessler reviewed (commented):
I think it's fine to demonstrate the prefit technique but in the context of kafe2 I think we should recommend the use of either a Poisson likelihood or its Gaussian approximation (depending on whether additional y errors are needed).

cverstege commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,166 @@
+We will especially look at two scenarios that can arise from having only small amounts of data.
+This is, for example, a problem in the search for new rare processes in high-energy physics.
+Assume we are looking for a Gaussian signal peak, free from background for simplicity
+(the treatment of a signal over background is explained in example 03_SpluBfit.py).
```
```suggestion
(the treatment of a signal over background is explained in example 03_SplusBfit.py).
```

cverstege commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,166 @@
+binedges = np.array([-2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2])
+
+#Now naively perform a default XYFit
+x_data = binmids = np.mean([binedges[:-1], binedges[1:]], axis=0) #use binmids as x values
```
I'd rather call it `bincenters` than `binmids`

cverstege commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,166 @@
+
+
+"""
+This first fit should give you warnings about the cost function being evaluated as infinite, which comes from the empty bin.
```
```suggestion
This first fit raises warnings about the cost function being evaluated as infinite, which originates from the empty bin.
```

cverstege commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,166 @@
+
+"""
+This first fit should give you warnings about the cost function being evaluated as infinite, which comes from the empty bin.
+Also, the result of the fit should not return any good values. This is the result of the empty bin.
```
```suggestion
Also, the result of the fit does not return any good values. This is the result of the empty bin.
```

cverstege commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,166 @@
+This first fit should give you warnings about the cost function being evaluated as infinite, which comes from the empty bin.
+Also, the result of the fit should not return any good values. This is the result of the empty bin.
+
+It can be seen in the following that even if we combine bins in order to get rid of the empty bin,
```
```suggestion
It can be seen in the following that even if the bins are combined in order to get rid of the empty bin,
```

cverstege commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,166 @@
+binedges = np.array([-2, -1, -0.5, 0, 0.5, 1, 2])
+
+#Now again perform a default XYFit
+x_data = binmids = np.mean([binedges[:-1], binedges[1:]], axis=0) #use binmids as x values
```
`bincenters` instead of `binmids`

cverstege commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,166 @@
+#Now again perform a default XYFit
+x_data = binmids = np.mean([binedges[:-1], binedges[1:]], axis=0) #use binmids as x values
+y_data = bincounts/(np.sum(bincounts)*np.diff(binedges)) #use normalized histogram as y data
+x_error = np.diff(binedges)/2 #use half binwidht as x_error
```
```suggestion
x_error = np.diff(binedges)/2 #use half binwidth as x_error
```

cverstege commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,166 @@
+plt.show()
+
+"""
+The uncertainties on the result were already reduced slightly in comparison to the prefit. And by looking at the printed
```
```suggestion
The uncertainties on the result were already reduced slightly in comparison to the prefit. And by looking at the
```

cverstege reviewed (commented):
Just some quick comments. Also, formatting is missing. You can run the black formatter to automatically fix the issues.

jonas-kleck committed:
reformatting and reformulation of histogram fit example and documentation

jonas-kleck committed:
incorpoate feedback into histogram fit section section

jonas-kleck committed:
changed the example. The focus lies more on 'recreating' a HistFit using XYFit

jonas-kleck committed:
changed the data compatibility check for XYFits using the poisson NLL to accept also non integer binedges

jonas-kleck commented:
After talking to @JohannesGaessler, I changed the example a bit. It should show more, how to recreate the Histfit and what the HistFit does internally. I also added a short part about when to use the gaussian approximation cost function.

JohannesGaessler commented on doc/src/parts/user_guide.rst:
```diff
@@ -389,6 +389,29 @@ A typical dictionary returned by the :py:meth:`~.FitBase.do_fit` method looks li
+---------------
+
+In physics experiments data is frequently histogrammed in order to reduce the data to a manageable number of bins.
+*kafe2* provides a dedicated fitting class for histogram fits, intendet to be used when datapoints are obtained
```
```suggestion
*kafe2* provides a dedicated fitting class for histogram fits, intended to be used when datapoints are obtained
```

JohannesGaessler commented on doc/src/parts/user_guide.rst:
```diff
@@ -389,6 +389,29 @@ A typical dictionary returned by the :py:meth:`~.FitBase.do_fit` method looks li
+
+In physics experiments data is frequently histogrammed in order to reduce the data to a manageable number of bins.
+*kafe2* provides a dedicated fitting class for histogram fits, intendet to be used when datapoints are obtained
+from a random distribution. Especiallywhen large numbers of datapoints are present it is more efficient to treat
```
```suggestion
from a random distribution. Especially when large numbers of datapoints are present it is more efficient to treat
```

JohannesGaessler commented on doc/src/parts/user_guide.rst:
```diff
@@ -389,6 +389,29 @@ A typical dictionary returned by the :py:meth:`~.FitBase.do_fit` method looks li
+In physics experiments data is frequently histogrammed in order to reduce the data to a manageable number of bins.
+*kafe2* provides a dedicated fitting class for histogram fits, intendet to be used when datapoints are obtained
+from a random distribution. Especiallywhen large numbers of datapoints are present it is more efficient to treat
+the data as a histogram. To perform a histogram fit, the datapoints have to be filled into a :py:obj:`HistContainer`.
```
```suggestion
the data as a histogram. To perform a histogram fit using raw data, the datapoints have to be filled into a :py:obj:`HistContainer`.
```

JohannesGaessler commented on doc/src/parts/user_guide.rst:
```diff
@@ -389,6 +389,29 @@ A typical dictionary returned by the :py:meth:`~.FitBase.do_fit` method looks li
+    hist_fit.do_fit()
+    
+By default it is assumed that the model function for a `HistFit` object is a probability density function normalised to 1.
+As a consequence the bin contents are also being normalized to 1.
```
```suggestion
By default it is assumed that the model function for a `HistFit` object is a probability density function normalized to 1.
As a consequence the bin contents are also being normalized to 1.
```

I don't particularly care which spelling convention we use but we should be consistent.

JohannesGaessler commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,162 @@
+HistFit class rather than the XYFit class.
+
+While it is in principle possible to perform such fits correctly using XYFit,
+more care must be taken to avoid unusable or biased results.
```
```suggestion
more care must be taken to avoid unusable or biased fit results.
```

JohannesGaessler commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,162 @@
+This example shows common problems that can occur when using XYFit and how to fix them.
+HistFit handles these problems automatically.
+
+We will try to build the kafe2 HistFit by hand using an XYFit. In general Histogram Fits are useful,
```
```suggestion
We will try to build the equivalent of the kafe2 HistFit by hand using an XYFit. In general histogram fits are useful
```

JohannesGaessler commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,162 @@
+
+We will try to build the kafe2 HistFit by hand using an XYFit. In general Histogram Fits are useful,
+when large amounts of data are obtained, since the binning of a histogram reduces the
+amount of computation.
```
```suggestion
when handling large amounts of data. The computational complexity of e.g. the chi2 cost function scales with the number of inputs cubed (when considering correlations). So reducing the raw data to a comparatively small number of bins is crucial to make the fit computationally feasible.
```

JohannesGaessler commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,162 @@
+The default XYFit has some problems that have to be addressed, when histogrammed data is
+processed. First of all, when using an XYFit, the data is passed to the Fit object in a
+XYContainer. This container does not automatically fill the datapoints in bins. So this first step
+has to be done manually. In the following, 30 datapoints, sampled from a normal distribution, are used.
```
I think we should just start with something like "The XYFit does not have any built-in functionality for transforming raw data into a histogram. So if we want to use it we need to do this step manually."

JohannesGaessler commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,162 @@
+    ]
+)
+
+binedges = np.array([-2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2])
```
```suggestion
bin_edges = np.array([-2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2])
```

I think it would make sense to put underscores between words for better legibility.

JohannesGaessler commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,162 @@
+Plot1.plot()
+plt.show()
+"""
+When performing the fit, errors like "The cost function was evaluated as infinite" appear in the output. 
+Furthermore, when looking at the plot of our fit result, it is clear that this fit didn't return the result we expected.
+The starting values for the fit are returned as best fit value, and the uncertainties are reported as NaN.
+What happened? The problem arises because Poisson uncertainties were assumed for the bin counts.
+If a bin is empty, the uncertainty is treated as zero by the fit. Thus the model function is forced to pass this datapoint
+exactly or the cost function will be infinite.
+This occures because the XYFit uses a χ²-cost function by default, which is only valid 
+for Gaussian uncertainties, but not in the case of Poisson uncertainties. To fix this, the
+cost function of the fit is changed to a Poisson negative log-likelhoood (NLL).
```
I think it would make sense to print a warning when using a cost function that needs errors and some of the errors are also == 0.

JohannesGaessler commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,162 @@
+
+Another subtlety is the definition of the y_data: So far, simply the midpoint of each bin was used. This is only
+a linear approximation of the behaviour of the model function between the bin edges. The HistFit class of kafe2
+on the other hand uses "Simpsons rule", a method to approximate the behaviour quadratically for more accuracy.
```
```suggestion
on the other hand uses "Simpson's rule", a method to approximate the behaviour quadratically for more accuracy.
```

JohannesGaessler commented on examples/009_histogram_fit/04_pitfalls.py:
```diff
@@ -0,0 +1,162 @@
+on the other hand uses "Simpsons rule", a method to approximate the behaviour quadratically for more accuracy.
+The most accurate albeit computationally expensive method would be to integrate the model function over each bin.
+
+The implementation of Simpsons rule in our procedure using a XYFit will not be done here,
```
```suggestion
The implementation of Simpson's rule in our procedure using a XYFit will not be done here,
```

JohannesGaessler reviewed (commented):
I think the example should be more clear that the "Guassian approximation" in the beginning and the "Gaussian approximation" cost function do different things: one is using the square root of the data, the other one is using the square root of the model.

The behavior that we should aim for is to use the Poisson likelihood cost function by default and to automatically switch to the Gaussian approximation if the user specifies Gaussian errors (I think I already implemented this).

jonas-kleck committed:
revised histogram fit section

jonas-kleck committed:
improved explanations in example

jonas-kleck committed:
edited warning message for zero errors in chi2 cost function
